File: 程序员面试金典/java/src/_08_03_MagicIndexLcci.java
/**
 * 第一遍：2020/07/31周五 ✅
 * 第二遍：2020/08/01周六 ✅
 * 第三遍：2020/07/29周三
 * 第四遍：2020/07/05周日
 */
class _08_03_MagicIndexLcci {
    public int findMagicIndex(int[] nums) {
        //b[i] = a[i] - i 不具有单调性，所以无法使用二分
        if (nums == null) {
            return -1;
        }
        int i = 0;
        while (i < nums.length) {
            if (nums[i] == i) {
                return i;
            }
            if (i < nums[i]) {
                i = nums[i];
            } else {
                i++;
            }
        }
        return -1;
    }
}

File: 程序员面试金典/kotlin/src/_08_03_MagicIndexLcci.kt
fun findMagicIndex(nums: IntArray): Int {
    //b[i] = a[i] - i 不具有单调性，故无法使用二分法
    var i = 0
    var n = nums.size
    while (i < n) {
        if (nums[i] == i) {
            return i
        }
        if (i < nums[i]) {
            i = nums[i]
        } else {
            i++
        }
    }
    return -1
}